File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> st;
        for(auto curr: asteroids){
            bool alive = true;

            while(alive && curr<0 && !st.empty() && st.top()>0){
                //upcoming astroid is larger
                int prev = st.top();
              

                if(prev<abs(curr)){
                    st.pop();
                }else if(prev==abs(curr)){
                    st.pop();
                    alive = false;
                }else{
                    alive = false;
                }
            }

            if(alive){
                st.push(curr);
            }
        }

        vector<int> ans;
        while(!st.empty()){
            ans.push_back(st.top()); st.pop();
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }
};

File: 0735-asteroid-collision/0735-asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        ans = []
        for i in range(len(asteroids)):
            curr = asteroids[i]
            alive = True
            while (alive and (len(stack) > 0) and (stack[-1] > 0) and (curr < 0)):
                # if both are same
                if stack[-1] == abs(curr):
                    stack.pop()
                    alive = False
                elif stack[-1] < abs(curr):
                    stack.pop()
                else:
                    alive = False

            if alive == True:
                stack.append(curr)

        while len(stack) > 0:
            ans.append(stack[-1])
            stack.pop()

        ans.reverse()

        return ans
